servers/src/mining/stratumserver.rs

```diff
@@ -742,69 +812,71 @@ impl StratumServer {
-					&self.chain,
-					&self.tx_pool,
-					self.verifier_cache.clone(),
-					self.current_key_id.clone(),
-					wallet_listener_url,
-				);
-				self.current_difficulty =
-					(new_block.header.total_difficulty() - head.total_difficulty).to_num();
-				self.current_key_id = block_fees.key_id();
-				current_hash = latest_hash;
-				// set the minimum acceptable share difficulty for this block
-				self.minimum_share_difficulty = cmp::min(
-					self.config.minimum_share_difficulty,
-					self.current_difficulty,
-				);
-				// set a new deadline for rebuilding with fresh transactions
-				deadline = Utc::now().timestamp() + attempt_time_per_block as i64;
-
 				{
-					let mut stratum_stats = stratum_stats.write();
+					let mut current_block_versions = self.current_block_versions.write();
+					let mut wallet_listener_url: Option<String> = None;
+					if !self.config.burn_reward {
+						wallet_listener_url = Some(self.config.wallet_listener_url.clone());
+					}
+					// If this is a new block, clear the current_block version history
+					if current_hash != latest_hash {
+						current_block_versions.clear();
+					}
+					// Build the new block (version)
+					let (new_block, block_fees) = mine_block::get_block(
+						&self.chain,
+						&self.tx_pool,
+						self.verifier_cache.clone(),
+						self.current_key_id.read().clone(),
+						wallet_listener_url,
+					);
+					{
+						let mut current_difficulty = self.current_difficulty.write();
+						*current_difficulty =
+							(new_block.header.total_difficulty() - head.total_difficulty).to_num();
+					}
+					{
+						let mut current_key_id = self.current_key_id.write();
+						*current_key_id = block_fees.key_id();
+					}
+					current_hash = latest_hash;
+					{
+						// set the minimum acceptable share difficulty for this block
+						let mut minimum_share_difficulty = self.minimum_share_difficulty.write();
+						*minimum_share_difficulty = cmp::min(
+							self.config.minimum_share_difficulty,
+							*self.current_difficulty.read(),
+						);
+					}
+					// set a new deadline for rebuilding with fresh transactions
+					deadline = Utc::now().timestamp() + attempt_time_per_block as i64;
+
+					let mut stratum_stats = self.stratum_stats.write();
 					stratum_stats.block_height = new_block.header.height;
-					stratum_stats.network_difficulty = self.current_difficulty;
+					stratum_stats.network_difficulty = *self.current_difficulty.read();
+
+					// Add this new block version to our current block map
+					current_block_versions.push(new_block);
 				}
-				// Add this new block version to our current block map
-				self.current_block_versions.push(new_block);
 				// Send this job to all connected workers
 				self.broadcast_job();
 			}
 
-			// Handle any messages from the workers
-			self.handle_rpc_requests(&mut stratum_stats.clone());
-
 			// sleep before restarting loop
-			thread::sleep(Duration::from_micros(1));
+			thread::sleep(Duration::from_millis(50));
```

Comment:
Why we increased sleep time?

Reply:
That's funny. We didn't increased it, it was like that when this branch was created,  and because I renamed a file git couldn't figure out what'd happened during rebase.

Reply:
Isn't 1 microsecond too small?  Wouldn't it be infinite loop eating out the whole CPU core?

Reply:
Agree